File: UICollectionView-Usage/UICollectionView-Usage/ViewController.m
//
//  ViewController.m
//  UICollectionView-Usage
//
//  Created by tzsoft on 2018/4/2.
//  Copyright © 2018年 胡小羊. All rights reserved.
//

#import "ViewController.h"

@interface ViewController ()<UICollectionViewDelegate,UICollectionViewDataSource,UICollectionViewDelegateFlowLayout>
@property (nonatomic , strong)UICollectionView *collectionView;
@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.view.backgroundColor = [UIColor lightGrayColor];
    [self.view addSubview:self.collectionView];
}

- (UICollectionView *)collectionView{
    if (!_collectionView) {
        //通过UICollectionViewFlowLayout类决定滚动方向
        UICollectionViewFlowLayout *layout = [[UICollectionViewFlowLayout alloc]init];
        //设置滑动方向
        layout.scrollDirection = UICollectionViewScrollDirectionVertical;
        layout.minimumLineSpacing = 10;
        
        _collectionView = [[UICollectionView alloc]initWithFrame:self.view.bounds collectionViewLayout:layout];
        _collectionView.delegate = self;
        _collectionView.dataSource = self;
        
        //注册cell类
        [_collectionView registerClass:[UICollectionViewCell class] forCellWithReuseIdentifier:@"UICollectionViewCell"];
    }
    return _collectionView;
}
#pragma mark - UICollectionViewDelegate,UICollectionViewDataSource
- (NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView{
    return 2;
}
- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section{
    return 5;
}
- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath{
    UICollectionViewCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:@"UICollectionViewCell" forIndexPath:indexPath];
    cell.backgroundColor = [UIColor redColor];
    return cell;
}

#pragma mark - UICollectionViewDelegateFlowLayout
// item的尺寸
- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath{
    return CGSizeMake(100, 100);
}
//设置距离屏幕上左下右的距离
- (UIEdgeInsets)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout insetForSectionAtIndex:(NSInteger)section{
    return UIEdgeInsetsMake(10, 10, 10, 10);
}
//设置header的大小
- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout referenceSizeForHeaderInSection:(NSInteger)section {
    return CGSizeMake(60, 60);
}
@end

File: Swift/基本语法/第二部分/02-02处理异常的练习/02-02处理异常的练习/ViewController.swift
//
//  ViewController.swift
//  02-02处理异常的练习
//
//  Created by tzsoft on 2018/3/7.
//  Copyright © 2018年 胡小羊. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        
        // 正则表达式
        //1、创建正则表达式的规则
        let pattern = "abc"
        //2、创建正则表达式对象
        
        // 方式一：try方式
        do{
            let regex = try NSRegularExpression(pattern: pattern, options: .caseInsensitive)
        }catch{
            
        }
        
        // 方式二：try？方式
        // 若有异常，则返回nil；没异常则返回可选类型，使用时要注意解包
        let regex2 = try? NSRegularExpression(pattern: pattern, options: .caseInsensitive)
        
        // 方式二：try!方式 （危险的方式）,表示肯定没有异常，返回一个确定类型；若是出现异常会出现崩溃
        let regex3 = try! NSRegularExpression(pattern: pattern, options: .caseInsensitive)
    }
}

